libs/gl-client/src/lnurl/pay/mod.rs

```diff
@@ -0,0 +1,251 @@
+fn parse_invoice(invoice_str: &str) -> Result<Invoice> {
+    Invoice::from_str(&invoice_str).map_err(|e| anyhow!(format!("Failed to parse invoice: {}", e)))
+}
+
+// Function to decode and parse the lnurl into a URL
+fn decode_and_parse_lnurl(lnurl: &str) -> Result<String> {
+    let (_hrp, data, _variant) =
+        bech32::decode(lnurl).map_err(|e| anyhow!("Failed to decode lnurl: {}", e))?;
+
+    let vec = Vec::<u8>::from_base32(&data)
+        .map_err(|e| anyhow!("Failed to base32 decode data: {}", e))?;
+
+    let url = String::from_utf8(vec).map_err(|e| anyhow!("Failed to convert to utf-8: {}", e))?;
+    Ok(url)
+}
+
+// Function to extract the description from the lnurl pay request response
+fn extract_description(lnurl_pay_request_response: &PayRequestResponse) -> Result<String> {
+    let mut description = String::new();
+
+    let serialized_metadata = lnurl_pay_request_response.metadata.clone();
+
+    let deserialized_metadata: Vec<Vec<String>> =
+        serde_json::from_str(&serialized_metadata.to_owned())
+            .map_err(|e| anyhow!("Failed to deserialize metadata: {}", e))?;
+
+    for metadata in deserialized_metadata {
+        if metadata[0] == "text/plain" {
+            description = metadata[1].clone();
+        }
+    }
+
+    ensure!(!description.is_empty(), "No description found");
+
+    Ok(description)
+}
+
+// Function to build the callback URL based on lnurl pay request response and amount
+fn build_callback_url(
+    lnurl_pay_request_response: &models::PayRequestResponse,
+    amount: u64,
+) -> Result<String> {
+
+    let mut url = Url::parse(&lnurl_pay_request_response.callback)?;
+    url.query_pairs_mut().append_pair("amount", &amount.to_string());
+    Ok(url.to_string())
+}
+
+// Validates the pay request response for expected values
+fn validate_pay_request_response(
+    lnurl_pay_request_response: &PayRequestResponse,
+    amount: u64,
+) -> Result<()> {
+    if amount < lnurl_pay_request_response.min_sendable {
+        return Err(anyhow!(
+            "Amount must be {} or greater",
+            lnurl_pay_request_response.min_sendable
+        ));
+    }
+
+    if amount > lnurl_pay_request_response.max_sendable {
+        return Err(anyhow!(
+            "Amount must be {} or less",
+            lnurl_pay_request_response.max_sendable
+        ));
+    }
+
+    if lnurl_pay_request_response.tag != "payRequest" {
+        return Err(anyhow!("Expected tag to say 'payRequest'"));
+    }
+
+    Ok(())
+}
+
+// Validates the invoice on the pay request's callback response
+fn validate_invoice_from_callback_response(
+    invoice: &Invoice,
+    amount: u64,
+    lnurl_pay_request: PayRequestResponse,
+) -> Result<()> {
+    ensure!(invoice.amount_milli_satoshis().unwrap_or_default() == amount ,
+        format!("Amount found in invoice was not equal to the amount found in the original request\nRequest amount: {}\nInvoice amount:{:?}", amount, invoice.amount_milli_satoshis().unwrap())
```

Comment:
The `ensure!` macro does support string formatting, as far as I understand, so you don't need to nest a `format!()` call into it.

Reply:
👍